src/libs/filterArrayByMatch.ts

```diff
@@ -0,0 +1,327 @@
+    let matchingInOrderCharCount = 0;
+    let charNumber = 0;
+    function findMatchingCharacter(matchChar: string, string: string, index: number) {
+        for (let j = index; j < string.length; j++) {
+            if (string[j] === matchChar) {
+                matchingInOrderCharCount += 1;
+                return j + 1;
+            }
+        }
+        return -1;
+    }
+
+    function getRanking(spread: number) {
+        const spreadPercentage = 1 / spread;
+        const inOrderPercentage = matchingInOrderCharCount / stringToRank.length;
+        const ranking = rankings.MATCHES + inOrderPercentage * spreadPercentage;
+
+        return ranking as Ranking;
+    }
+
+    const firstIndex = findMatchingCharacter(stringToRank[0], testString, 0);
+
+    if (firstIndex < 0) {
+        return rankings.NO_MATCH;
+    }
+
+    charNumber = firstIndex;
+    for (let i = 1; i < stringToRank.length; i++) {
+        const matchChar = stringToRank[i];
+        charNumber = findMatchingCharacter(matchChar, testString, charNumber);
+        const found = charNumber > -1;
+
+        if (!found) {
+            return rankings.NO_MATCH;
+        }
+    }
+
+    const spread = charNumber - firstIndex;
+    return getRanking(spread);
+}
+
+/**
+ * Gives a rankings score based on how well the two strings match.
+ * @param testString - the string to test against
+ * @param stringToRank - the string to rank
+ * @returns the ranking for how well stringToRank matches testString
+ */
+function getMatchRanking(testString: string, stringToRank: string): Ranking {
+    // too long
+    if (stringToRank.length > testString.length) {
+        return rankings.NO_MATCH;
+    }
+
+    // case sensitive equals
+    if (testString === stringToRank) {
+        return rankings.CASE_SENSITIVE_EQUAL;
+    }
+
+    // Lower casing before further comparison
+    const lowercaseTestString = testString.toLowerCase();
+    const lowercaseStringToRank = stringToRank.toLowerCase();
+
+    // case insensitive equals
+    if (lowercaseTestString === lowercaseStringToRank) {
+        return rankings.EQUAL;
+    }
+
+    // starts with
+    if (lowercaseTestString.startsWith(lowercaseStringToRank)) {
+        return rankings.STARTS_WITH;
+    }
+
+    // word starts with
+    if (lowercaseTestString.includes(` ${lowercaseStringToRank}`)) {
+        return rankings.WORD_STARTS_WITH;
+    }
+
+    // contains
+    if (lowercaseTestString.includes(lowercaseStringToRank)) {
+        return rankings.CONTAINS;
+    }
+    if (lowercaseStringToRank.length === 1) {
+        return rankings.NO_MATCH;
+    }
+
+    // acronym
+    if (getAcronym(lowercaseTestString).includes(lowercaseStringToRank)) {
+        return rankings.ACRONYM;
+    }
+
+    // will return a number between rankings.MATCHES and rankings.MATCHES + 1 depending  on how close of a match it is.
+    return getClosenessRanking(lowercaseTestString, lowercaseStringToRank);
```

Comment:
the difference between `getMatchRanking` and `getClosenessRanking` ranking is not immediately clear. Given that `getClosenessRanking` is only used once, let's just inline it here.